tools/clang/tools/dxcompiler/dxclinker.cpp

```diff
@@ -229,13 +318,74 @@ HRESULT STDMETHODCALLTYPE DxcLinker::Link(
 
     // Attach libraries.
     bool bSuccess = true;
-    for (unsigned i = 0; i < libCount; i++) {
+    const DeserializedDxilCompilerVersion *cur_version = nullptr;
+    const DeserializedDxilCompilerVersion *first_version = nullptr;
+
+    std::string cur_lib_name;
+    std::string first_lib_name;
+
+    for (UINT32 i = 0; i < libCount; i++) {
       CW2A pUtf8LibName(pLibNames[i], CP_UTF8);
       bSuccess &= m_pLinker->AttachLib(pUtf8LibName.m_psz);
+
+      cur_lib_name = std::string(pUtf8LibName);
+
+      // only libraries with compiler version parts are in the map
+      auto result = m_libNameToCompilerVersionPart.find(cur_lib_name);
+
+      if (result != m_libNameToCompilerVersionPart.end()) {
+        cur_version = ((*result).second);
+      } else {
+        UINT32 valMajor, valMinor;
+        dxcutil::GetValidatorVersion(&valMajor, &valMinor);
+        bool bValidatorAtLeast_1_8 =
+            DXIL::CompareVersions(valMajor, valMinor, 1, 8) >= 0;
+        if (bValidatorAtLeast_1_8) {
+          std::string errorMsg =
+              "error: Cannot link library with no compiler version part.\n";
```

Comment:
From our discussions, it sounded like perhaps the version was coming from the compiler not the validator. If that's the case, checking the validator version doesn't seem right. Is it?

Regardless, this is going to fall through for the first iteration anyway and produce additional errors anyway. Why is this special error here?

Reply:
There are 3 cases that can happen:

1. The compiler must link a set of libraries that each have dxil containers that contain compiler version parts.
2. The compiler must link a set of libraries, where not every library has a compiler version part in the dxil container.
3. The compiler must link a set of libraries, where no dxil container has a compiler version part.

There should be no error for case 1, assuming all the parts are the same. There should be an error for case 2, because a compiler version part existing for one library implies the library with the non-existent part is different. However, it isn't obvious how to move forward in case 3, because we may want to treat a set of libraries with no compiler version parts as equal, and we may later want to treat them as unequal. I introduced the validator check because I thought it was logical that linking a set of libraries without compiler version parts depends on the version of the validator, and the check may set a framework for adjusting the version later. (Adjusting the version threshold would prevent users with older validator versions from seeing errors). But, we may also consider disallowing case 3 and emitting an error unconditionally.

Reply:
I think basing the expectation of a version part on the validator version of each library in `RegisterLibrary()` could make sense, since the version part is not legal in current released validators.  Let's call this the "version part expected depends on lib validator version" approach

Given that, we could say that adding libraries without an *expected* version part is an error.  Then we can go ahead and link libraries where none of them have a version part, because that's what's required for the linker to keep working for existing users and an updated compiler for retail scenarios.

The downside to failing in `RegisterLibrary()` on a missing *expected* version part is that it has no error message output, only an HRESULT.  An unexpected new `E_INVALIDARG` is not likely to go over well, especially without any clear indication as to what went wrong.  This means the checking and error reporting would likely need to remain in `Link()`.

Another downside to this ("version part expected depends on lib validator version") approach would be blocking attempted use of older compiler versions to compile a library, when working around a bug for instance.  If that were to occur after a validator that supports the version part, you would be unable to link that library.  

The current approach in this PR of using the current validator version to determine whether it's legal to link libraries without version parts (let's call it "version parts expected depends on linker validator version") will prevent linking of libraries compiled (or validated) with a compiler (or validator) version that did not support the version part, if a validator version that supports the version part is available at link time.  I'm not sure that's the right approach either, but there is a setting for validator version, and if we respect that, we should be able to offer a way for scenarios to continue to work in this intermediate time frame.

I'm also concerned about the scenario where pre-built libraries are shipped to link with user code, which could be the case for DXR scenarios.

Ultimately, this compiler version usage is the lazy, brittle way to avoid having to carefully and explicitly version the lib ABI/conventions.  I still fear turning it on without an opt-out, breaking users, and expecting to get that feedback and respond by adding an opt-out mechanism if necessary.  It's likely users will find and stick with a workaround of their own if they can find one, rather than give us the feedback and wait for us to do something to unblock them.  We don't want them doing that, because it may block users from adopting newer compilers, features, and validators, or at least add friction here.

Reply:
I think the best approach may be to simply allow the case where no libraries being linked have a compiler version part.

That will probably be the least problematic and blocking.  You would still get errors when compiling new libraries with the part and trying to link them with other libraries without the part, or with a different version.  That would be the new default behavior for compilation going forward, so it's only in legacy/corner cases that you would pass through this no version part path, and I think it's acceptable to allow it.